File: GAMSampleApp/ios/GAMSampleApp/HeaderBidding/RewardedAd/POBRNRewardedAd.m
#import "POBRNRewardedAd.h"
#import "POBRNAdHelper.h"
#import <OpenWrapHandlerDFP/OpenWrapHandlerDFP.h>

@interface POBRNRewardedAd() <POBRewardedAdDelegate>
@property (nonatomic, strong) POBRewardedAd *rewardedAd;
/** Emitter used for emitting the global event for the interstitial ad. */
@property (nonatomic, nullable, weak) POBRNEventEmitter *eventEmitter;
@end

@implementation POBRNRewardedAd

- (instancetype)initWithEventEmitter:(POBRNEventEmitter *)eventEmitter
                         publisherId:(NSString *)publisherId
                           profileId:(NSNumber *)profileId
                          owAdUnitId:(NSString *)owAdUnitId
                         gamAdUnitId:(NSString *)gamAdUnitId {
    self = [super init];
    if (self) {
        _eventEmitter = eventEmitter;

        // The event handler is used to make the ad request to GAM after OW SDK
        // has received the bid.
        // The event handler requires gamAdUnitId for making the ad ad request to GAM.
        DFPRewardedEventHandler *handler = [[DFPRewardedEventHandler alloc] initWithAdUnitId:gamAdUnitId];

        // Config block for setting custom targeting or extra parameters on GAM request.
        handler.configBlock = ^(GAMRequest *request, POBBid *bid) {
            // Here, you have a provision to set the custom targeting to ad server using the
            // bid parameter as illustrated below:
            // NSDictionary *gamCustomTargeting = [[NSDictionary alloc] initWithObjectsAndKeys:@"value1", @"key1", nil];
            // [request setCustomTargeting:gamCustomTargeting];
        };

        // Pass the instance of event handler along with other parameters while creating
        // instance of POBRewardedAd.
        self.rewardedAd = [POBRewardedAd rewardedAdWithPublisherId:publisherId
                                                         profileId:profileId
                                                          adUnitId:owAdUnitId
                                                      eventHandler:handler];
        self.rewardedAd.delegate = self;
    }

    return self;
}

#pragma mark - POBRNFullScreenAd

- (void)destroyAd {
    // Destroy rewarded ad.
    self.rewardedAd.delegate = nil;
    self.rewardedAd = nil;
}

- (BOOL)isReady {
    // Whether rewarded ad is ready.
    return self.rewardedAd.isReady;
}

- (void)loadAd {
    // Load rewarded ad.
    [self.rewardedAd loadAd];
}

- (void)showAd {
    // Show rewarded ad.
    [self.rewardedAd showFromViewController:[POBRNAdHelper topViewController]];
}

#pragma mark - POBRewardedAdDelegate

// Notifies the delegate that an rewarded ad has been received successfully.
-(void) rewardedAdDidReceiveAd:(POBRewardedAd *)rewardedAd {
    [self sendEvent:POBRN_EVENT_AD_RECEIVED];
}

// Notifies the delegate of an error encountered while loading an ad.
-(void) rewardedAd:(POBRewardedAd *)rewardedAd didFailToReceiveAdWithError:(NSError *)error {
    [self sendEvent:POBRN_EVENT_FAILED_TO_LOAD_AD error:error];
}

// Notifies the delegate of an error encountered while showing an ad.
-(void) rewardedAd:(POBRewardedAd *)rewardedAd didFailToShowAdWithError:(NSError *)error {
    [self sendEvent:POBRN_EVENT_FAILED_TO_SHOW_AD error:error];
}

// Notifies the delegate that the rewarded ad will be presented as a modal on top of the current view controller.
-(void) rewardedAdWillPresentAd:(POBRewardedAd *)rewardedAd {
    // Not applicable
}

// Notifies the delegate that the rewarded ad is presented as a modal on top of the current view controller.
-(void) rewardedAdDidPresentAd:(POBRewardedAd *)rewardedAd {
    [self sendEvent:POBRN_EVENT_AD_OPENED];
}

// Notifies the delegate that the rewarded ad has been animated off the screen.
-(void) rewardedAdDidDismissAd:(POBRewardedAd *)rewardedAd {
    [self sendEvent:POBRN_EVENT_AD_CLOSED];
}

// Notifies the delegate of ad click
-(void) rewardedAdDidClickAd:(POBRewardedAd *)rewardedAd {
    [self sendEvent:POBRN_EVENT_DID_CLICK_AD];
}

// Notifies the delegate that a user interaction will open another app (e.g. App Store),
// leaving the current app. To handle user clicks that open the
// landing page URL in the internal browser, use 'RewardedAdDidClickAd:' instead.
-(void) rewardedAdWillLeaveApplication:(POBRewardedAd *)rewardedAd {
    [self sendEvent:POBRN_EVENT_WILL_LEAVE_APP];
}

// Notifies the delegate of an ad expiration. After this callback,
// this 'POBRewardedAd' instance is marked as invalid & will not be shown.
-(void) rewardedAdDidExpireAd:(POBRewardedAd *)rewardedAd {
    [self sendEvent:POBRN_EVENT_AD_EXPIRED];
}

// Notifies the delegate that a user will be rewarded once the ad is completely viewed.
-(void) rewardedAd:(POBRewardedAd *)rewardedAd shouldReward:(POBReward *)reward {
    [self.eventEmitter emitEventWithEmbeddedEventName:POBRN_EVENT_RECEIVE_REWARD reward:reward];
}

#pragma mark - Private methods

// Send event
- (void)sendEvent:(NSString *)eventName {
    [self sendEvent:eventName error:nil];
}

// Send event with error
- (void)sendEvent:(NSString *)eventName error:(nullable NSError *)error {
    [self.eventEmitter emitEventWithEmbeddedEventName:eventName error:error];
}

@end

File: GAMSampleApp/android/app/src/main/java/com/pubmatic/sampleapp/rewarded/POBRNRewardedAd.kt
package com.pubmatic.sampleapp.rewarded

import androidx.annotation.MainThread
import com.facebook.react.bridge.ReactApplicationContext
import com.pubmatic.sampleapp.interfaces.POBRNFullScreenAd
import com.pubmatic.sampleapp.util.POBRNSDKConstant
import com.pubmatic.sampleapp.util.POBRNSDKUtils
import com.pubmatic.sdk.common.POBError
import com.pubmatic.sdk.openwrap.core.POBReward
import com.pubmatic.sdk.openwrap.eventhandler.dfp.DFPRewardedEventHandler
import com.pubmatic.sdk.rewardedad.POBRewardedAd
import org.json.JSONObject

/**
 * This class demonstrates how to integrate OpenWrap SDK rewarded ads through GAM Header Bidding in React Native environments.
 * It follows the FullScreenAd protocol which is a wrapper on POBRewardedAd which helps you to load and show rewarded ads.
 * You can also add custom targeting to GAM by using config block and also use methods to facilitate your use case by going through OpenWrap SDK integration guide.
 * This class can be reused or customized as per the requirement in your React Native application.
 */
@MainThread
class POBRNRewardedAd(
    publisherId: String,
    profileId: Int,
    owAdUnitId: String,
    gamAdUnitId: String,
    var context: ReactApplicationContext
) : POBRNFullScreenAd {

    // Used to store the instance of [POBRewardedAd] for performing actions
    // such as, loadAd, showAd, etc.
    private var rewarded: POBRewardedAd? = null

    /**
     *  Init block to initialize rewarded ad object
     */
    init {
        context.currentActivity?.let { activity ->
            // The event handler is used to make the ad request to GAM after OW SDK
            // has received the bid.
            // The event handler requires two paramters -
            // - The reference of the current activity for showing the ad.
            // - gamAdUnitId for making the ad ad request to GAM.
            val eventHandler = DFPRewardedEventHandler(activity, gamAdUnitId)

            eventHandler.setConfigListener { requestBuilder, bid ->
                // Here, you have a provision to set the custom targeting to ad server
                // Example:
                //    requestBuilder.addCustomTargeting("<key>", "<value>")
            }

            // Pass the instance of event handler along with other parameters while creating
            // instance of POBRewardedAd.
            rewarded = POBRewardedAd.getRewardedAd(
                context, publisherId, profileId, owAdUnitId,
                eventHandler
            )

            // Set OW SDK event listener.
            rewarded?.setListener(POBRewardedAdListener())
        }
    }

    override fun loadAd() {
        // Load rewarded ad.
        rewarded?.loadAd()
    }

    override fun isReady(): Boolean {
        // Whether rewarded ad is ready.
        rewarded?.let {
            return it.isReady
        }
        return false
    }

    override fun showAd() {
        // Show rewarded ad.
        rewarded?.show()
    }

    override fun destroyAd() {
        // Destroy rewarded ad.
        rewarded?.destroy()
        rewarded = null
    }

    /**
     * Internal class which implement POBRewardedAd Listener and communicates callbacks to React
     */
    inner class POBRewardedAdListener :
        POBRewardedAd.POBRewardedAdListener() {
        /**
         * Notifies the listener that an ad has been successfully loaded and rendered.
         */
        override fun onAdReceived(ad: POBRewardedAd) {

            POBRNSDKUtils.emitCallbackEvent(
                context,
                POBRNSDKConstant.REWARDED_AD_EVENT,
                POBRNSDKConstant.AD_RECEIVED_EVENT,
                null
            )
        }

        /**
         * Notifies the listener of an error encountered while loading an ad.
         */
        override fun onAdFailedToLoad(ad: POBRewardedAd, error: POBError) {

            POBRNSDKUtils.emitCallbackEvent(
                context,
                POBRNSDKConstant.REWARDED_AD_EVENT,
                POBRNSDKConstant.AD_FAILED_TO_LOAD_EVENT,
                POBRNSDKUtils.getErrorExtJSON(error)
            )
        }

        /**
         * Notifies the listener of an error encountered while rendering an ad.
         */
        override fun onAdFailedToShow(ad: POBRewardedAd, error: POBError) {
            POBRNSDKUtils.emitCallbackEvent(
                context,
                POBRNSDKConstant.REWARDED_AD_EVENT,
                POBRNSDKConstant.AD_FAILED_TO_SHOW_EVENT,
                POBRNSDKUtils.getErrorExtJSON(error)
            )
        }

        /**
         * Notifies that the user has clicked the ad view.
         */
        override fun onAdClicked(ad: POBRewardedAd) {
            POBRNSDKUtils.emitCallbackEvent(
                context,
                POBRNSDKConstant.REWARDED_AD_EVENT,
                POBRNSDKConstant.AD_CLICKED_EVENT,
                null
            )
        }

        /**
         * Notifies listener that the rewarded ad has open the ad on top of the current
         * view/screen, as a result of user interaction.
         */
        override fun onAdOpened(ad: POBRewardedAd) {

            POBRNSDKUtils.emitCallbackEvent(
                context,
                POBRNSDKConstant.REWARDED_AD_EVENT,
                POBRNSDKConstant.AD_OPENED_EVENT,
                null
            )
        }

        /**
         * Notifies that the rewarded ad has closed the ad on top of the current view.
         */
        override fun onAdClosed(ad: POBRewardedAd) {
            POBRNSDKUtils.emitCallbackEvent(
                context,
                POBRNSDKConstant.REWARDED_AD_EVENT,
                POBRNSDKConstant.AD_CLOSED_EVENT,
                null
            )
        }

        /**
         * Notifies the listener that an ad has been expired
         */
        override fun onAdExpired(ad: POBRewardedAd) {

            POBRNSDKUtils.emitCallbackEvent(
                context,
                POBRNSDKConstant.REWARDED_AD_EVENT,
                POBRNSDKConstant.AD_EXPIRED_EVENT,
                null
            )
        }

        /**
         * Notifies the listener whenever current app goes in the background due to user click
         */
        override fun onAppLeaving(ad: POBRewardedAd) {
            POBRNSDKUtils.emitCallbackEvent(
                context,
                POBRNSDKConstant.REWARDED_AD_EVENT,
                POBRNSDKConstant.APP_LEAVE_EVENT,
                null
            )
        }

        /**
         * Notifies that the rewarded ad has completed and user should be rewarded. It is called
         * when the Rewarded Ad playback is completed.
         */
        override fun onReceiveReward(ad: POBRewardedAd, reward: POBReward) {
            val ext = JSONObject()
            ext.put(POBRNSDKConstant.REWARD_CURRENCY_TYPE_KEY, reward.currencyType)
            ext.put(POBRNSDKConstant.REWARD_AMOUNT_KEY, reward.amount)
            POBRNSDKUtils.emitCallbackEvent(
                context,
                POBRNSDKConstant.REWARDED_AD_EVENT,
                POBRNSDKConstant.RECEIVE_REWARD_EVENT,
                ext
            )
        }
    }
}
